// @rsbuild/core/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  PLUGIN_CSS_NAME: () => import_constants.PLUGIN_CSS_NAME,
  PLUGIN_LESS_NAME: () => import_constants.PLUGIN_LESS_NAME,
  PLUGIN_SASS_NAME: () => import_constants.PLUGIN_SASS_NAME,
  PLUGIN_STYLUS_NAME: () => import_constants.PLUGIN_STYLUS_NAME,
  PLUGIN_SWC_NAME: () => import_constants.PLUGIN_SWC_NAME,
  createRsbuild: () => import_createRsbuild.createRsbuild,
  defineConfig: () => import_config.defineConfig,
  loadConfig: () => import_config.loadConfig,
  loadEnv: () => import_loadEnv.loadEnv,
  logger: () => import_shared.logger,
  mergeRsbuildConfig: () => import_mergeConfig.mergeRsbuildConfig,
  rspack: () => import_core.rspack,
  version: () => version
});
module.exports = __toCommonJS(src_exports);
var import_core = require("@rspack/core");
var import_loadEnv = require("./loadEnv");
var import_createRsbuild = require("./createRsbuild");
var import_config = require("./config");
var import_shared = require("@rsbuild/shared");
var import_mergeConfig = require("./mergeConfig");
var import_constants = require("./constants");
const version = "0.6.15";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PLUGIN_CSS_NAME,
  PLUGIN_LESS_NAME,
  PLUGIN_SASS_NAME,
  PLUGIN_STYLUS_NAME,
  PLUGIN_SWC_NAME,
  createRsbuild,
  defineConfig,
  loadConfig,
  loadEnv,
  logger,
  mergeRsbuildConfig,
  rspack,
  version
});
